Issue#91 Cooperation CRUD

alexorlenko opened the pull request:
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)


## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

Add Cooperation Entity to application and CRUD operation

## Summary of change

Add new Entity

## Testing approach

-

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

Commit:
Issue#91. Add Cooperation, House and Address Entity to application and yaml file.

Commit:
Issue#91. Fix yaml file and add CooperationDto and HouseDto.

Commit:
Issue#91. Add CooperationRepository and HouseRepository.

Commit:
Issue#91. Add CooperationService and HouseService.

Commit:
Issue#91. Add CooperationServiceImpl and HouseServiceImpl.

Commit:
Issue#91. Add other method to CooperationServiceImpl.

Commit:
Issue#91. Delete not use annotation.

Commit:
Issue#91: created CooperationApiImpl.java

Commit:
Issue#91: implemented create/getById cooperation methods

Commit:
Merge branch 'dev' of https://github.com/ita-social-projects/Home into Issue#91-Cooperation-CRUD

 Conflicts:
	home-open-api/src/main/resources/yaml/openapi.yaml

Commit:
Issue#91: implemented getHouseById/updateHouse methods

Commit:
Issue#91: filled up permissions in database

Commit:
Issue#91: implemented delete CRUD methods

Commit:
Issue#91: implemented query methods

Commit:
Issue#91: passed checkstyle plugins

denis-litvinov committed:
Merge branch 'dev' into Issue#91-Cooperation-CRUD

Commit:
Merge branch 'dev' of https://github.com/ita-social-projects/Home into Issue#91-Cooperation-CRUD

 Conflicts:
	home-data-migration/src/main/resources/db/changelog/0.0.1/_cumulative.xml

Commit:
Issue#91: created api test classes

Commit:
Merge remote-tracking branch 'origin/Issue#91-Cooperation-CRUD' into Issue#91-Cooperation-CRUD

Commit:
Issue#91: returned Cooperation and House

Commit:
Issue#91: merged Contacts from dev

Commit:
Merge remote-tracking branch 'origin/Issue#91-Cooperation-CRUD' into Issue#91-Cooperation-CRUD

Commit:
Issue#91:

- implemented CooperationApiIT
- implemented QueryCooperationIT
- implemented HouseApiIT

Commit:
Issue#91:

- implemented QueryHouseIT
- fixed rsql mappings in CoopApiImpl

Commit:
Revert "Issue#91: passed checkstyle plugins"

This reverts commit b8d097b8

Commit:
Issue#91: passed checkstyle

Commit:
Issue#91: pulled Jacoco from dev

Commit:
Issue#91: fixed some mistakes

Commit:
Issue#91: fixed code smells and bug

kh0ma commented on home-api-tests/pom.xml:
```diff
@@ -341,5 +341,10 @@
             <artifactId>commons-text</artifactId>
             <version>${commons-text-version}</version>
         </dependency>
+        <dependency>
```
Why do we need it? We use junit5 and its dependency has been already declared.

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,138 @@
+class CooperationApiIT {
+
+    private final CooperationApi cooperationApi = new CooperationApi(ApiClientUtil.getClient());
+
```
Remove extra nextline

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,138 @@
+        ApiResponse<Void> response = cooperationApi.removeCooperationWithHttpInfo(readCoop.getId());
+
+        assertEquals(Response.Status.NO_CONTENT.getStatusCode(), response.getStatusCode());
+        assertThatExceptionOfType(ApiException.class)
```
Also please check that the proper status code is returned after this check.

denis-litvinov replied:
added this method 

".withStackTraceContaining("\"response_code\":404")"

denis-litvinov replied:
added this method: 
.withMessage(createExceptionMessage(exception)

exception has 404 response code

kh0ma replied:
Please use `.matches((actual) -> exception.getCode() == 404);` insted.

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,138 @@
+        assertEquals(update.getUsreo(), updated.getUsreo());
+        assertEquals(update.getAddress(), updated.getAddress());
+    }
+
```
Extra line

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+
+}
```
Are you going to add any tests for negative scenarios? Payload validations, business logic etc.

denis-litvinov replied:
Added.

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/QueryCooperationIT.java:
```diff
@@ -0,0 +1,145 @@
+    @Test
+    void getAllCooperationFilteredBy() throws ApiException {
+        CreateCooperation expected = createCooperation();
+        expected.setName("testedByFilters");
```
Is the name of the cooperation unique? Please make the name more random in order to avoid errors in case of running this test twice on the same seed.

kh0ma replied:
Is it resolved?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,183 @@
+                                     @Min(0) @Max(10) Integer pageSize,
+                                     String sort,
+                                     String filter,
+                                     String name,
```
`id` must be included as a query parameter

denis-litvinov replied:
In BaseQuery class we have this field id, but it's of String type, maybe i should change it to Long type to change all the queryBuilders?

kh0ma replied:
You should add it to the spec for this operation.

kh0ma replied:
Ok. I see it. Resolved.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+        Map<String, String> filterMap = new HashMap<>();
+
+        String quantityFlatValue = quantityFlat == null ? null : String.valueOf(quantityFlat);
```
Can we just do `quantityFlat.toString()` over `String.valueOf(quantityFlat)`?

kh0ma replied:
Is it resolved?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,183 @@
+        Map<String, String> filterMap = new HashMap<>();
+
+        String quantityFlatValue = quantityFlat == null ? null : String.valueOf(quantityFlat);
+        String adjoiningAreaValue = adjoiningArea == null ? null : String.valueOf(adjoiningArea);
```
The same as:

> Can we just do quantityFlat.toString() over String.valueOf(quantityFlat)?

kh0ma replied:
Is it resolved?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,183 @@
+        String houseAreaValue = houseArea == null ? null : houseArea.toString();
+        String houseIdValue = houseId == null ? null : houseId.toString();
+
+        filterMap.put("cooperation.id", cooperationId.toString());
```
filterMap.put("cooperation_id", cooperationId.toString());

kh0ma replied:
Is it resolved?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+        filterMap.put("cooperation.id", cooperationId.toString());
+        filterMap.put("id", houseIdValue);
+        filterMap.put("quantityFlat", quantityFlatValue);
```
### The key of this map must be the name of the query parameter from open api specification.

kh0ma replied:
Is it resolved?

kh0ma replied:
It isn't resolved. Please pay attention that the query parameter name in the spec is `quantity_flat`. So the put into this map must look like:
`filterMap.put("quantity_flat", quantityFlatValue);`

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/CooperationRSQLEndpointConfig.java:
```diff
@@ -0,0 +1,36 @@
+    public Map<CooperationQueryParamEnum, String> getMappings() {
+        EnumMap<CooperationQueryParamEnum, String> map = new EnumMap<>(CooperationQueryParamEnum.class);
+
+        map.put(CooperationQueryParamEnum.NAME, "name");
```
Please pay attention that the value of this map must be the query name from openapi spec.

kh0ma replied:
Is it resolved?

kh0ma replied:
It seems like yes.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/HouseRSQLEndpointConfig.java:
```diff
@@ -0,0 +1,36 @@
+    public Map<HouseQueryParamEnum, String> getMappings() {
+        EnumMap<HouseQueryParamEnum, String> map = new EnumMap<>(HouseQueryParamEnum.class);
+
+        map.put(HouseQueryParamEnum.COOPERATION_ID, "cooperation_id");
```
Please pay attention that the value of this map must be the query name from openapi spec.

kh0ma replied:
Is it resolved?

kh0ma replied:
It seems like Yes.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Cooperation.java:
```diff
@@ -15,8 +21,34 @@
+    @Column(name = "iban")
+    private String iban;
+
+    @Column(name = "registerDate")
```
I see that column names have either camel or snake case. Let's decide on one approach. Please highlight it on team meeting and propagate outcomes as part of you changes.

kh0ma commented on home-data/src/main/resources/application.properties:
```diff
@@ -2,4 +2,4 @@ spring.datasource.url=${DATASOURCE_URL:jdbc:postgresql://localhost:5433/postgres
 spring.datasource.password=${DATASOURCE_PASSWORD:password}
 spring.datasource.driver-class-name=org.postgresql.Driver
-spring.jpa.hibernate.ddl-auto=validate
+spring.jpa.hibernate.ddl-auto=validate
```
Please do not forget to add the last empty line to the file.

kh0ma replied:
Is it resolved?

kh0ma replied:
I can be wrong but github still shows diff here. Unresolve this.

denis-litvinov replied:
Done.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -17,6 +17,12 @@ tags:
   - name: contact
     description: Contact management operations
+  - name: cooperation
+    description: Cooperation managment
```
Typo: -> management

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -232,7 +238,7 @@ paths:
             application/json:
               schema:
-                $ref: '#/components/schemas/ReadUser'
+                  $ref: '#/components/schemas/ReadUser'
```
?

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successful created
```
Typo?

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+    post:
+      tags:
+        - cooperation
+      summary: "Create cooperation"
```
Why do we need Quotation marks?

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+          name: page_size
+          schema:
+            type: integer
+            minimum: 0
```
"0" really?

denis-litvinov replied:
I changed it to 1 in queryCooperation and queryHouse, now it works ok, but when we put 0 in payload we get 400 error response. We all have similar 0 in queryUsers, queryNews, queryContacts. Should i change it either?

kh0ma replied:
Yep

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+          name: filter
+          schema:
+            type: string
+            example: "some example"
```
?

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+          schema:
+            type: string
+            default: id,asc
+            example: "some example"
```
?

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,433 @@ paths:
+      parameters:
+        - name: "cooperation_id"
+          in: "path"
+          description: "Use 1 for testing."
```
# Please work on the descriptions, summaries etc. I provided detailed review and suggestion under https://github.com/ita-social-projects/Home/pull/179 See comments on openapi spec. Please propagate it to your changes.

kh0ma replied:
Any updates for this request?

denis-litvinov replied:
Made all the updates.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1441,172 @@ components:
+            $ref: "#/components/schemas/ReadHouse"
+    UpdateCooperation:
+      type: object
+      properties:
```
As we use it for PUT all the required fields from creating also must be here.

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1441,172 @@ components:
+          example: 12
+        address:
+          $ref: '#/components/schemas/Address'
+#        cooperationId:
```
Please remove commented lines.

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1441,172 @@ components:
+        adjoining_area:
+          type: integer
+          example: 12
+#        cooperationId:
```
Please remove commented lines.

kh0ma replied:
Is it resolved?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1441,172 @@ components:
+#        cooperationId:
+#          type: integer
+#          example: 2
+    UpdateHouse:
```
As we use it for PUT all the required fields from creating must be here.

kh0ma replied:
Is it resolved?

VadymSokorenko committed:
Issue#91: Minor changes.

 - Replaced strings with string.format in CooperationServiceImpl.
 - Changed method naming in openapi.yaml.

tex1988 commented on home-data-migration/src/main/resources/db/changelog/0.0.1/Issue91.xml:
```diff
@@ -0,0 +1,177 @@
+
+</databaseChangeLog>
+
+
```
Maybe it wil be better to use loadData change type and separated csv file with data

kh0ma replied:
I think loadData is mostly applicable for test data as we have to set runOnChange to “true” to allow rerunning changeset whenever CSV file is changed. We must not allow to do it.

denis-litvinov replied:
Done.

Commit:
Issue#91: corrected requested changes after review

Commit:
Issue#91: added tests for invalid input data for

- CooperationApiIT
- QueryCooperationIT
- HouseApiIT (addresses)

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,186 @@
+    @Test
+    void deleteCooperationTest() throws ApiException {
+        ReadCooperation readCoop = cooperationApi.createCooperation(createCooperation());
+        ApiException exception = new ApiException(404, "Can't find cooperation with given ID: " + readCoop.getId());
```
Why do we need to build this object? Can we just assert the message?

denis-litvinov replied:
Done.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,187 @@
+    @Autowired
+    private HouseService houseService;
+
+
```
Extra lines

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+        return Response.status(Response.Status.OK).entity(readCoop).build();
+    }
+
```
Extra line

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,187 @@
+        filterMap.put("cooperation_id", cooperationId.toString());
+        filterMap.put("id", houseIdValue);
+        filterMap.put("quantityFlat", quantityFlatValue);
+        filterMap.put("adjoiningArea", adjoiningAreaValue);
```
Same

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,187 @@
+        filterMap.put("id", houseIdValue);
+        filterMap.put("quantityFlat", quantityFlatValue);
+        filterMap.put("adjoiningArea", adjoiningAreaValue);
+        filterMap.put("houseArea", houseAreaValue);
```
Same

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Cooperation.java:
```diff
@@ -15,8 +21,34 @@
 @Table(name = "cooperations")
 @SequenceGenerator(name = "sequence", sequenceName = "cooperations_sequence")
-public class Cooperation extends BaseEntity implements Serializable {
+public class Cooperation extends BaseEntity implements Serializable{
```
Do we need ` implements Serializable` here? It is already in BaseEntity.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+    private static final String NOT_FOUND_COOPERATION = "Can't find cooperation with given ID:";
+
+    private static final String FORMAT = "%s %d";
```
What's going on here? 

Why cannot we combine `NOT_FOUND_COOPERATION` with `FORMAT`?

So we'll have single constant with a message format:
```java
private static final String NOT_FOUND_COOPERATION_FORMAT = "Can't find cooperation with given ID: %d";
```

Commit:
Issue#91: corrected 2 iterated requested changes:

- openapi.yaml summaries, descriptions
- tests

Commit:
Issue#91: fixed two code smells

Abhai2016 commented on home-data-migration/src/main/resources/db/changelog/0.0.1/Issue91.xml:
```diff
@@ -0,0 +1,177 @@
+                <constraints nullable="false"/>
+            </column>
+        </addColumn>
+        <addColumn tableName="houses">
```
Can we do it in one tag "addColumn"? I don't have the answer so I'm just curious if we could do it.

denis-litvinov replied:
Seems it worked!

Abhai2016 commented on home-data-migration/src/main/resources/db/changelog/0.0.1/Issue91.xml:
```diff
@@ -0,0 +1,177 @@
+        <insert dbms="postgresql"
+                schemaName="public"
+                tableName="permissions">
+            <column name="id" value="10"/>
```
Do we have to hardcore id's value? Because we have sequences now so there's probebly a way how to use it here by using  tag "valueComputed" instead of "value" for example.

denis-litvinov replied:
We agreed to create additional ticket for this

Abhai2016 commented on home-data-migration/src/main/resources/db/changelog/0.0.1/Issue91.xml:
```diff
@@ -0,0 +1,177 @@
+        </addColumn>
+    </changeSet>
+
+    <changeSet id="Issue91.006" author="denis-litvinov">
```
I guess the first we have to do is create sequences and then already make other changes to use those sequences for example.

denis-litvinov replied:
But it doesn't prevent right now cause we make changes using already created sequances by Dima's task before. Here created sequences are not used right now in changes

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successfully created
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+            example:  "A9856738"
+      responses:
+        '200':
+          description: successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+            example: 1
+      responses:
+        "204":
+          description: Successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successfully created
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+            example: 20
+      responses:
+        '200':
+          description: Successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+            example: 1
+      responses:
+        '200':
+          description: Successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+        required: true
+      responses:
+        '200':
+          description: Successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -669,6 +675,438 @@ paths:
+            example: 1
+      responses:
+        "204":
+          description: Successful operation
```
Success*

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1446,176 @@ components:
         phone:
           type: string
           example: "+380966617305"
+
```
Do we need this extra line?

denis-litvinov replied:
Updated

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1006,3 +1446,176 @@ components:
+          type: integer
+          example: 12
+        address:
+          $ref: '#/components/schemas/Address'
```
Please add a new line in the end of the file.

denis-litvinov replied:
Updated

Abhai2016 commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,184 @@
+
+    @PreAuthorize(CREATE_COOPERATION_PERMISSION)
+    @Override
+    public Response createCooperation(@Valid CreateCooperation createCooperation) {
```
Got confused about @Valid. We added it in new api but I can't see it in old implementations. Why do we need it and if we do why didn't we add it in other apis?

kh0ma replied:
It looks like it happens on implementing the method by using IntelliJ IDEA "implement" action and this method parameter annotation was taken from the CooperationApi interface. But AFAIK it is redundant here as Jersey will take care of it via looking at the whole hierarchy and apply it automatically.

denis-litvinov replied:
Deleted @Valid annotation

Abhai2016 commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,184 @@
+
+    @PreAuthorize(GET_HOUSES_PERMISSION)
+    @Override
+    public Response queryHouse(@Min(1) Long cooperationId,
```
I can see we decided to not do separated api for house and it's fine because it works fine with contacts for user as well. But what will be with flats, they will be in cooperation api as well?

kh0ma replied:
Do you mean keeping it in separate controllers? It is a gap or unknown capability from our generator. As for now, It generates only the root level controller and places all sub-resources there.

But I agree that we have to consider its refactoring.

Abhai2016 commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,184 @@
+
+        filterMap.put("cooperation_id", cooperationId.toString());
+        filterMap.put("id", houseIdValue);
+        filterMap.put("quantity_flat", quantityFlatValue);
```
I guess you can try get string already in the method or use null-safe operator like quantityFlat?.toString()

kh0ma replied:
Groovy jokes?)))

Abhai2016 commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Address.java:
```diff
@@ -0,0 +1,39 @@
+@SequenceGenerator(name = "sequence", sequenceName = "addresses_sequence")
+public class Address extends BaseEntity {
+
+    @Column(name = "region")
```
As I know you can to not specify name in @Column if it matches with column name

kh0ma replied:
As for this, we have to decide on always use defaults or keep everything strict. One pitfall of this default behavior that when we rename the field name we can break it since nothing is changed in DB schema.

Abhai2016 commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/dto/HouseDto.java:
```diff
@@ -0,0 +1,18 @@
+
+    private Integer adjoiningArea;
+
+    private Address address;
```
Don't we need a cooperation in HouseDto?

denis-litvinov replied:
As i see, we do not use it in any operations in service layers

Abhai2016 commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    public CooperationDto getCooperationById(Long id) {
+        Cooperation toGet = cooperationRepository.findById(id).filter(Cooperation::getEnabled)
+            .orElseThrow(() -> new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id)));
+        List<House> houseList = houseRepository.findHousesByCooperationId(toGet.getId());
```
I'm not sure if we need this because if we did everything right then hibernate has to get list of houses automatically .
Because when we get user with all his contacts we don't make another query for exampel so we can check how it's implemented there.

kh0ma replied:
I reckon that this story will prevent us from such situations https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/202

Abhai2016 commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+        Cooperation toDelete = cooperationRepository.findById(id).filter(Cooperation::getEnabled)
+            .orElseThrow(() -> new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id)));
+        toDelete.setEnabled(false);
+        cooperationRepository.save(toDelete);
```
What about houses which were in this cooperation, don't we have to delete them as well or change cooperation for them?

kh0ma replied:
Good question. It depends on how we are going to use children after disabling parents. To avoid huge cascade updates let's disable the only parents and in the future keep the children invisible in the system if a parent is disabled.

Abhai2016 commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/HouseServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+        House toGet = houseRepository.findById(id).filter(House::getEnabled)
+            .orElseThrow(() -> new NotFoundHomeException(String.format(HOUSE_WITH_ID_NOT_FOUND, id)));
+        if (!toGet.getCooperation().getId().equals(coopId)) {
+            throw new NotFoundHomeException(String.format(HOUSE_WITH_ID_NOT_FOUND, id));
```
I think you have to separate messages for the user when he wants to get the house which doesn't exist with not_found message and when he wants to get house which exists but not in presented cooperation with not_allowe message of something.

kh0ma replied:
Why cannot we combine it using proper predicate? e.g. we have `/parents/5/children/13` then the query will be `select * from children where id = '13' and parent_id = '5'` In this case, we will have only one message pointing that the assessed resource is not found.

kh0ma commented on home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/cooperations/CooperationApiIT.java:
```diff
@@ -0,0 +1,185 @@
+
+    @Test
+    void createCooperationInvalidNameTest() {
+        ApiException exception = new ApiException(400, "Parameter `name` is invalid - size must be between 1 and 50 signs.");
```
Why do we need it? Can we just match the message? Please look at all similar places and fix them.

Also I reckon that method `createExceptionMessage(exception)` must be removed after it.

denis-litvinov replied:
Updated

Commit:
Issue#91: corrected Abhai requested changes

Commit:
Merge branch 'dev' of https://github.com/ita-social-projects/Home into Issue#91-Cooperation-CRUD

 Conflicts:
	home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/query/CooperationQueryConfig.java
	home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/query/HouseQueryConfig.java

likeRewca commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CooperationApiImpl.java:
```diff
@@ -0,0 +1,184 @@
+    @Override
+    public Response createCooperation(@Valid CreateCooperation createCooperation) {
+        CooperationDto createCoopDto = mapper.convert(createCooperation, CooperationDto.class);
+        CooperationDto readCoopDto = cooperationService.createCooperation(createCoopDto);
```
It would be appropriate to return ReadCooperation object in the createCooperation method, so we will remove unnecessary mappings and remove 2 lines

denis-litvinov replied:
We will mix layers if we'll remove ReadCooperation object to service layers. Presentation layer and service layer must not know about each other

likeRewca commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    @Transactional
+    @Override
+    public CooperationDto updateCooperation(Long id, CooperationDto updateCooperationDto) {
+        Optional<Cooperation> optional = cooperationRepository.findById(id).filter(Cooperation::getEnabled);
```
Optional have method orElseThrow(() ->new NotFoundHomeException()). With this method u can delete first "if"

denis-litvinov replied:
Updated!

Commit:
Issue#91: implemented 190 task changes with query

Commit:
Issue#91: removed optional object from update methods in all services

denis-litvinov committed:
Merge branch 'dev' into Issue#91-Cooperation-CRUD

Commit:
Merge branch 'dev' of https://github.com/ita-social-projects/Home into Issue#91-Cooperation-CRUD

Commit:
Issue#91: fixed code smells

Commit:
Merge remote-tracking branch 'origin/Issue#91-Cooperation-CRUD' into Issue#91-Cooperation-CRUD
